### PR TITLE
video_core: Reimplement the buffer cache

### PR DESCRIPTION
**Pull request by ReinUsesLisp**
Reimplement the buffer cache using cached bindings and page level
granularity for modification tracking. This also drops the usage of
shared pointers and virtual functions from the cache.

- Bindings are cached, allowing to skip work when the game changes few
  bits between draws.
- OpenGL Assembly shaders no longer copy when a region has been modified
  from the GPU to emulate constant buffers, instead GL_EXT_memory_object
  is used to alias sub-buffers within the same allocation.
- OpenGL Assembly shaders stream constant buffer data using
  glProgramBufferParametersIuivNV, from NV_parameter_buffer_object. In
  theory this should save one hash table resolve inside the driver
  compared to glBufferSubData.
- A new OpenGL stream buffer is implemented based on fences for drivers
  that are not Nvidia's proprietary, due to their low performance on
  partial glBufferSubData calls synchronized with 3D rendering (that
  some games use a lot).
- Most optimizations are shared between APIs now, allowing Vulkan to
  cache more bindings than before, skipping unnecessary work.

This commit adds the necessary infrastructure to use Vulkan object from
OpenGL. Overall, it improves performance and fixes some bugs present on
the old cache. There are still some edge cases hit by some games that
harm performance on some vendors, this are planned to be fixed in later
commits.

Regression tracking:
- [x] Fire Emblem: Three Houses crashing before showing anything on Vulkan.
- [x] Xenoblade Vulkan won't load fonts.
- [x] Fire Emblem Warriors black screen at boot.
- [x] Animal Crossing: New Horizons black screen.
- [x] Fire Emblem: Three Houses crashes on OpenGL assembly shaders.
- [x] Age of Calamity doesn't render on GLSL.
- [x] Bravely Default 2 Demo black screen.
- [x] Bravely Default 2 crashes on Vulkan.
- [x] Final Fantasy X crashes at boot on Vulkan.
- [x] Dragonball Fighters Z doesn't render on GLSL.
- [x] Hollow Knight has pixelated fog on GLSL.
- [x] Hollow Knight crashes on Vulkan.
- [x] Link's Awakening boot screen doesn't render on GLSL.
- [x] Oninaki vertex explosions.
- [x] Shin Megami Tensei 3 has broken graphics on GLSL.
- [x] ~~Mario and Rabbids has graphical issues.~~ Happens on mainline Vulkan
- [x] Several games softlock on Vulkan.
- [x] Astral Chain can crash/softlock on some cutscenes.
- [x] ~~Final Fantasy Maxima crashes on Vulkan.~~ Unrelated to this PR, fixed in #5796.
- [ ] Crash Nitro Fueled no longer boots with high accuracy on Intel.
- [x] Pokémon Let's go Eevee is slower on Intel.
- [x] Astral Chain crashes on Vulkan when trying to use a null descriptor as index buffer.
- [ ] Link's Awakening has animation issues with chain chomps on Pascal and older generations on Vulkan with high GPU accuracy.

### CODE REVIEW

**Comment by FearlessTobi** [comment]
Should fix https://github.com/yuzu-emu/yuzu/issues/3918.

**Comment by lat9nq** [comment]
Fixes #3951

**Comment by gidoly** [comment]
Fix #4923

**Comment by dawnbomb** [comment]
a massive amount of games have tons of problems. According to hearsay through mod GoldenX86 to rodrigo the GPU dev, this is occuring on pascal based GPU's. Severity ranges from major to oh my fucking god how did this happen tier, and affects a huge amount of games.  some crash, others just get examples like images below

https://imgur.com/a/ZOBVSgR
https://cdn.discordapp.com/attachments/476077620350418964/800213492291338270/unknown.png
https://cdn.discordapp.com/attachments/476077620350418964/800213520749953115/unknown.png

it affects VULKAN and OPENGL, but not equally. they can have diffrent new problems for the same game, many games have a multitude of problems all with the same game and same graphics engine. Below are Debug logs of only some affected games (it takes time). It is worth noting that sometimes, i get to see graphics live 'un-load' infront of my eyes. i have attempted to record this, but it basically crashes my entire PC if i do. I do have one attempted recording with shareX that didn't crash my PC, and got shareX's error log spitting out massive memory errors, i put the sharex log in a pastebin here for sake of clarity: https://pastebin.com/qDEU1wSB

To start with, altho not reported offically on the wiki, the wiki for 3H is outdated, and before BCR it was 100% crash free and running well. It just takes like 350 hours to do all the content in the game so i didn't have enough time to report, but i checked all known crashes on the wiki and they were NOT crashing before BCR, please don't say 3H was known unplayable as a scapegoat, i can show proof its fine before if anyone cares. anyway moving on...

Yuzu Vulkan crashes before loading anything (Load save -> Crash)
https://www.mediafire.com/file/8u4u3dggv5n2d1q/yuzu_log%25283HVulkan%2529.txt/file

Xenoblade Vulkan won't load fonts
https://www.mediafire.com/file/zu09yomg1bunt6c/yuzu_log%2528Xenoblade%2529.txt/file

FE Warriors black screen on boot (audio still plays)
https://www.mediafire.com/file/qvbem7mfta19xe9/yuzu_log%2528Warriors%2529.txt/file

Animal crossing black screen, big icons and dialogue boxes appear, and sound plays
https://www.mediafire.com/file/mg8ewn2qx5x02b4/yuzu_log%2528AnimalCrossing%2529.txt/file
https://cdn.discordapp.com/attachments/476077620350418964/800226153197469716/unknown.png

3H OpenGL Crash
https://www.mediafire.com/file/9aeytuxca07gtb8/yuzu_log%25283HOpenGL%2529.txt/file
This one seems unusual, usually openGL crashes as sothis unloads infront of my eyes, this one i got the idea to record on my phone sense local recording crash, and i uploaded to youtube.
https://www.youtube.com/watch?v=ycm18gVKnu0&feature=youtu.be

Also reported by others who won't post logs, is splatoon 2, and links awakening, but there is definatly more. 

Edit: Here is one i really wanted to show off, a video of what appears to be the live unload of assets.
https://www.mediafire.com/file/hub7nzlpg5703gc/yuzu_log%25283HUnloadingOpenGL%2529.txt/file
https://www.youtube.com/watch?v=EViMWsu2j-A&feature=youtu.be

**Comment by gidoly** [comment]
i tested various game and i manage to see various issue
1. age of calamity not render on glsl
2.bravely default 2 final demo title screen not render on glsl


3. bravely default 2 final demo crash on vulkan and glasm api
4. dragonball fighters z not render on glsl (black screen only selection menu thingy able to see it)
5.  final fantasy 10 remaster crashes on vulkan after start new game (specifically x-1)
6. hollow knight lighting pixelate on glsl

how it should look

7.hollow knight crashes on vulkan after pass first cutscene of game
8.zelda links awakening intro video not work properly on glsl (black out and only show final section of vid
9.oninaki vertex explosion on all 3 api
10.shin megami tensei 3 remaster graphic render broken on glsl

how it should look like

11. graphic issue and frame drop issue on mario rabbid ingdom on glsl (i think graphic issue all appear on 3 api)


i will add other if i manage to find it.

**Comment by lightgod87** [comment]
Good day,

I have to report a regression in terms of performance on Pokémon Sword and Shield. Before BCR, the game is running smoothly.

[yuzu_log_debug_EA1326.zip](https://github.com/yuzu-emu/yuzu/files/5827186/yuzu_log_debug_EA1326.zip)

But after BCR (in EA1335 on this case), the performance became sluggish for some reason.

[yuzu_log_debug_EA1335.zip](https://github.com/yuzu-emu/yuzu/files/5827188/yuzu_log_debug_EA1335.zip)

Keep in mind these specs:

Intel Core i7-8700
Nvidia Geforce GTX 1070 Ti (with the driver version 461.09)
16 GBs DDR4
Windows 10

That's all for now.

**Comment by BSoDGamingYT** [comment]
> Good day,
> 
> I have to report a regression in terms of performance on Pokémon Sword and Shield. Before BCR, the game is running smoothly.
> 
> [yuzu_log_debug_EA1326.zip](https://github.com/yuzu-emu/yuzu/files/5827186/yuzu_log_debug_EA1326.zip)
> 
> But after BCR (in EA1335 on this case), the performance became sluggish for some reason.
> 
> [yuzu_log_debug_EA1335.zip](https://github.com/yuzu-emu/yuzu/files/5827188/yuzu_log_debug_EA1335.zip)
> 
> Keep in mind these specs:
> 
> Intel Core i7-8700
> Nvidia Geforce GTX 1070 Ti (with the driver version 461.09)
> 16 GBs DDR4
> Windows 10
> 
> That's all for now.
> 
> EDIT: Same thing happens in EA1336.
> 
> [yuzu_log_debug_EA1336.zip](https://github.com/yuzu-emu/yuzu/files/5827204/yuzu_log_debug_EA1336.zip)



> Good day,
> 
> I have to report a regression in terms of performance on Pokémon Sword and Shield. Before BCR, the game is running smoothly.
> 
> [yuzu_log_debug_EA1326.zip](https://github.com/yuzu-emu/yuzu/files/5827186/yuzu_log_debug_EA1326.zip)
> 
> But after BCR (in EA1335 on this case), the performance became sluggish for some reason.
> 
> [yuzu_log_debug_EA1335.zip](https://github.com/yuzu-emu/yuzu/files/5827188/yuzu_log_debug_EA1335.zip)
> 
> Keep in mind these specs:
> 
> Intel Core i7-8700
> Nvidia Geforce GTX 1070 Ti (with the driver version 461.09)
> 16 GBs DDR4
> Windows 10
> 
> That's all for now.
> 
> EDIT: Same thing happens in EA1336.
> 
> [yuzu_log_debug_EA1336.zip](https://github.com/yuzu-emu/yuzu/files/5827204/yuzu_log_debug_EA1336.zip)

Please enable High GPU accuracy

File: src/video_core/rasterizer_interface.h [hunk 1]
```diff
@@ -7,6 +7,7 @@
 #include <atomic>
 #include <functional>
 #include <optional>
+#include <span>
```
Comment (ameerj) [inline src/video_core/rasterizer_interface.h:10]: unnecessary include?

File: src/video_core/renderer_opengl/gl_rasterizer.h [hunk 2]
```diff
@@ -152,6 +133,9 @@ class RasterizerOpenGL : public VideoCore::RasterizerAccelerated {
     /// Configures images in a compute shader.
     void SetupComputeImages(const Shader* shader);
 
+    /// Syncs state to match guest's
```
Comment (ameerj) [inline src/video_core/renderer_opengl/gl_rasterizer.h:136]:
```suggestion
    /// Syncs state to match the guest state
```

**Review by ameerj — COMMENTED** [review]
few minor nits

File: src/video_core/rasterizer_interface.h (hunk 1 above)
Comment (ReinUsesLisp) [reply src/video_core/rasterizer_interface.h:10]: Yup! I used to have a function to push constants directly into host GPU memory, but it was too slow. I'll remove it

**Comment by gidoly** [comment]
i found new regression
on astral chain all 3 api crash or freeze before loading render of this cutscene
log:
[log.zip](https://github.com/yuzu-emu/yuzu/files/5833969/log.zip)

save:

[01007300020FA000.zip](https://github.com/yuzu-emu/yuzu/files/5833973/01007300020FA000.zip)

**Comment by IGCIT** [comment]
intel hw gen11 (tested on linux, mesa 20.3.3, opengl)

fixes:
 * major slowdown in Pokken - now fully playable
 * major glitches in Crash nitro fueled - now fully playable (only minor graphics anomalies in main map)
 * minor glitches in Sky force anniversary - now fully playable
 * black grass in Pokemon mystery dungeon dx in first dungeon

regressions:
 * Crash nitro fueled - no longer booting when gpu accuracy is set to high (working before)
 * probably more but i'm not sure these commits are the cause so i will open a separate issue

**Comment by v1993** [comment]
Results of testing (KUbuntu 20.10, Nvidia GTX 960, driver 460.32.03):

* Kirby Star Allies: some performance degradation in some places and slight improvement in others.
* Cadence of Hyrule: extreme performance improvements (game no longer GPU bound), but softlocks (partially? some animations keep running and sound loops over the same short fragment slowly advancing forwards) after a while in dungeons.

**Comment by Banjo-Kazooie-fan** [comment]
https://github.com/yuzu-emu/yuzu/issues/5769

**Comment by IGCIT** [comment]
new regression found for intel

pokemon let's go, major slowdown, from 30fps to 4fps with bcr

**Comment by ReinUsesLisp** [comment]
Edited OP to have all known regressions. Please comment here which ones are fixed, or ping me on discord.

**Comment by ReinUsesLisp** [comment]
@IGCIT On Pokémon Let's go and Crash Team Racing I softlock when booting those games on OpenGL. On the other hand they run fine on Vulkan.
Does Crash Team Racing's `svcBreak` happen to you on OpenGL or Vulkan? Same for Pokémon?

**Comment by JRBdatascience** [comment]
CPU AMD 4500u 6c/6t
GPU: Integrated Vega 6 - Adrenaline 21.1.1 Drivers (Also happens with 20.12.1)
RAM: 12 GB
OS: Windows 10 20H2
Build: EA 1357


Dragon Quest XI crashes on OpenGL with “Debug <Critical> video_core\texture_cache\image_view_base.cpp:ImageViewBase:29: Assertion Failed! Image view format R32G32B32A32_FLOAT is incompatible with image format R32G32_FLOAT”

Probably Duplicate of #5775 but Pokemon Shield is stuck on Launching Game with "Game: 0 FPS Frame: Nan ms" shown in Vulkan. This happens with GPU accuracy set to normal or high. I included a log for a short test where I waited 1 minute for something to happen, and a long test where I waited 5 mins. The logs ended up with similar sizes though. I can play the game in OpenGL.

[Shield (Short Test).txt](https://github.com/yuzu-emu/yuzu/files/5855002/Shield.Short.Test.txt)

[Shield (Long Test).txt](https://github.com/yuzu-emu/yuzu/files/5855005/Shield.Long.Test.txt)

**Comment by dawnbomb** [comment]
Reports on regression fixes

FE warriors regressions were fixed. Some may report it still has crashes, however those crashes were from before BCR, the BCR specific crashes were all fixed as far as i can tell.
https://www.mediafire.com/file/urw528pipsuay8z/yuzu_logWarriorsCrash.txt/file
This is the log for it still crashing. I decided to include it here, even tho this is *not* a regression crash, because it is technically still new. Before BCR it would crash differently, now it crashes like this. To my knowledge, this crash hasn't been reported before, but it is unknown if it's directly because of BCR as it's new. I would be fine with it being ignored for now, but i wanted to report it anyway as it's still technically new.

Xenoblade font was fixed.

3 Houses OpenGL assembly shaders crash was... *probably* fixed.
However, what i reported before using vulkan was only because it happens randomly, and is annoying to recreate. It happens more often on vulkan, but it still happens on open GL.
Vulkan Log: https://www.mediafire.com/file/1jy52x5ul4bkukl/yuzu_log3H_Vulkan2.txt/file
Open GL Log: https://www.mediafire.com/file/ynxmui8nt3bgj0h/yuzu_log3HOpenGL_Crash2.txt/file

It is mainly easier however to test on vulkan, as openGL has crashes before BCR, while vulkan did not. So the openGL version of the crash could be scapegoated as a non-BCR crash, but the vulkan one is *defiantly* because of BCR. Still, it is the same crash on OpenGL or Vulkan, same conditions, same results.  I find the crash to happen more often when pressing new game, but that takes longer per attempt to trigger the crash to get to the first map.

Also, that checklist is really hard to read the boxes in dark mode :p

**Comment by IGCIT** [comment]
@ReinUsesLisp 

well, i don't know your gpu but current status on my intel iris plus 940/g7 is:

**Windows 10** (i dont play yuzu here that much due to broken drivers, but i did some tests now):
i'm now running latest driver 9168

**Opengl**:

well, things are not good here, all my games have issues with ogl, glitches and/or freezes.
ogl is completely broken in windows, i know that latest intel drivers (at least since 9126) broke ogl in multiple ways so i'm not sure this is 100% yuzu fault.
i mean master has less issues but other emulators, that had no changes, now have glitches with ogl, so...it is possible that bcr is unmasking some driver issues

anyway

**Pokemon let's go**:
this is actually something new to me, when i tried the game with driver 9126 and a 3-4 days old build it was fine, i mean too slow but no locks.

current status:
the game boots here and there are lots of glitches in main menu and some glitches in-game but i get mixed results.
sometime it runs for a few minutes, sometime black screen and sometime it freezes instantly when i load my save, but every time the game is still running, audio is still playing and i can interact with the game but i'm blind :P

when this happens, if i stop the game i cannot run any other game in yuzu and i get the message: `Your GPU may not support OpenGL, or you do not have the latest driver`

yuzu log says:
```
Frontend    yuzu/bootmanager.cpp:OpenGLShadedContext:142: Unable to create main openGL context
Frontend    yuzu/bootmanager.cpp:OpenGLShadedContext:157: Unable to create shared openGL context
Frontend    yuzu/bootmanager.cpp:OpenGLShadedContext:157: Unable to create shared openGL context
```

i have to close and reopen yuzu to fix this.

i see no svc break on my log.

pikagli



**Crash nitro**:
no svc break or softlocks.

game is 10fps (full speed on master)

major glitches are gone but there are new issues with brc, i see a transparent copy of myself on the left and a semi-transparent black square on the right.

if i start a race it freezes.
again, game is still running and if i stop it, no openGL error as described above.
in windows the game boots even with gpu accuracy set to high

crashnitro-intel-windows


**Vulkan**:

things are a bit better but yuzu crashes when booting some games

**Pokemon let's go**:
no glitches and freezes.

the game is a bit slow, 20-25 fps, 30 when less things on screen, but i don't know if this is supposed to be like that

also, we have an open issue: [vkCreateGraphicsPipelines on multiple threads](https://github.com/IGCIT/Intel-GPU-Community-Issue-Tracker-IGCIT/issues/8) that could be the reason to low performance in vulkan (i dont know if yuzu is affected by this)

**Crash nitro**:
full speed and no glitches, but yuzu crashes after a few minutes or if it does not crash when you boot the game, it crashes when you start a race,
log is always truncated and i see no svc break in what i have


i noticed now that logs are often truncated in windows but not in linux.


**ArchLinux: (where i play yuzu)**

**OpenGL**:

**Pokemon let's go**:
no glitches or freezes
the game is really slow, full speed on master

**Crash nitro**:
when gpu accuracy is set to high, yuzu crashes, instantly or after a few minutes.

here i have the svc break

bcr has fixed all the major glitches and this is how it looks now in main map (the black bands)

crashnitro-mesa



**Vulkan**:
NA

stuck at "loading" when booting every game with vulkan, probably unrelated to this pr?

```
[  35.330468] Kernel.SVC <Error> core/hle/kernel/svc.cpp:MapSharedMemory:1202: Address does not fit within the map region, addr=0x000000195D604000, siz
e=0x0000000001100000
[  35.330641] Kernel.SVC <Error> core/hle/kernel/svc.cpp:MapSharedMemory:1202: Address does not fit within the map region, addr=0x0000001C16204000, siz
e=0x0000000001100000
```

**Comment by ReinUsesLisp** [comment]
@IGCIT Thanks for detailed explanation! After fixing some issues on Vulkan, I'll look into the mesa GL issues with my UHD 630. I did some performance tuning for Breath of the Wild on that driver, but I didn't test the Pokémons.

**Comment by dawnbomb** [comment]
So, i'm not sure why all the 3H crashes got checkmarked, because i just posted logs yesterday that it still has regression crashes caused by BCR. Accident?

**Comment by JRBdatascience** [comment]
Update for EA 1360

I don't get the crash on OpenGL on Dragon Quest anymore. I get in game and it is a glitchy mess, but mainline 514 is the same.

I still get stuck on Launching with Pokemon Shield on Vulkan. OpenGL goes in game fine.

**Comment by goldenx86** [comment]
@dawnbomb Those crashes are confirmed to also happen in mainline.

**Comment by ReinUsesLisp** [comment]
@JRBdatascience managed to reproduce the issue. Investigating a fix.

**Comment by ReinUsesLisp** [comment]
855bcb0 fixes the issue. The next EA (not the one that is currently building, the next one) will have that issue fixed.

**Comment by dawnbomb** [comment]
@goldenx86 mainline? maybe, idk, but they defiantly do NOT happen before BCR for EA at least. it was defiantly a BCR regression. 
I guess i could test mainline for them as well.

**Comment by ReinUsesLisp** [comment]
@IGCIT Use high accuracy to fix the performance issue on Linux when playing Pokémon games.
Most of our time is spent queuing and popping GPU readbacks, and it's not an Intel exclusive issue.
image

**Comment by IGCIT** [comment]
ok, this fixed pokemon slowdown.
thanks for the clarification

**Comment by ameerj** [comment]
#5329 was fixed by this PR when it was first opened, but the regression is back when I tested on EA 1392.

**Comment by G-Spawn** [comment]
EA 1442. 
Another regression to note in Dead Cells. Affects title menu's and in game (missing text of oddities in game shown below). Assembly shaders being on or off seems to have a big impact on it.

image

image

Assembly shaders on ^
Corresponding log. 

[yuzu_log.txt](https://github.com/yuzu-emu/yuzu/files/5963684/yuzu_log.txt)


image

https://user-images.githubusercontent.com/23653025/107612814-135e6b00-6c15-11eb-9b14-cfcee0f4f316.mp4

^ Assembly off.
Corresponding log. 
[yuzu_log.txt](https://github.com/yuzu-emu/yuzu/files/5963707/yuzu_log.txt)


Hope the info provided helps!

**Comment by lat9nq** [comment]
In addition to the missing header errors that every other CI bot is reporting issues with, here are the errors that Clang 11 generated.
[bcr-errors.txt](https://github.com/yuzu-emu/yuzu/files/5975111/bcr-errors.txt)

**Review by bunnei — APPROVED** [review]

**Pull request merged**